### DISCUSSION TITLE
Welcome to the Packet Sender Forums!

### DISCUSSION THREAD

**Discussion by dannagle** [dannagle/PacketSender#215] [General]
## Welcome to the new Packet Sender Forums!
This is my 3rd attempt at creating forums. My previous 2 attempts were overrun by spammers. GitHub now has a new "Discussions" feature, and it looks very promising. GitHub has been hosting my code and issues for the past 5 years, and I've been pleased. Lets give these Forums another go, shall we? 3rd time is the charm.

Feel free to:
  * Ask general Packet Sender questions.
  * Ask _me_, or others, questions about anything software development related. e.g. Why did I do X and not Y? I love helping others learn. 
  * Ask for help.
  * Discuss networking.
  * Share cool Packet Sender ideas. You integrators frequently amaze me.
  * Chat with other users.
  * Basically, anything outside of bugs and feature requests may go to these forums.  I'm keeping it quite open (for now).
 
My hope is this space will provide a good area for discussion while helping me keep the Issues tab focused on whatever work that is needed by the project. 

So how are you using Packet Sender? Please introduce yourself.

**Comment by dannagle** [comment]
I'll go first.  My name is Dan Nagle. I wrote and open sourced Packet Sender on GitHub in 2015. 

At the time I started the project, I was working in the network AV industry. I needed a quick and simple way to **correctly** send custom UDP and TCP commands to devices to help test them. I also needed to **receive** unsolicited packets to test the device clients. Nothing out there did it quite right, or it did too complicated. Later on, I needed SSL support too. 

I eventually decided to create my own from scratch. I had a wishlist, and this list still guides the project:
- Packet Sender places a strong emphasis on being easy to use. 
- Packet Sender emphasizes compatibility across Windows and Mac. 
- Packet Sender can be ran without admin permissions and without any other dependencies. 

I want to be able to download, run, and train a new user of the tool in under 5 minutes. This is actually how Packet Sender Cloud was started -no jump drive needed.

I'm happy to answer any questions you may have.

**Comment by rwhitenet** [comment]
Hi Dan -
Looks like this is a new forum. Great. Question: is there any way to set the source port number on TCP packets?

Thanks, R.White

**Reply by dannagle** [reply]
Not via the desktop GUI. It always requests a dynamic port. The command line has a `--bind` option that can be used for TCP/SSL. 
```
PS C:\Program Files\PacketSender> ./packetsender -atw 500 --bind 5215 127.0.0.1 5000 "hello from CLI."
TCP (5215)://127.0.0.1:5000 68 65 6c 6c 6f 20 66 72 6f 6d 20 43 4c 49 2e

Response Time:16:22:10.851
Response HEX:48 65 6C 6C 6F 20 66 72 6F 6D 20 47 55 49
Response ASCII:Hello from GUI
PS C:\Program Files\PacketSender>
```

**Comment by johnathanschneider** [comment]
Hey Dan! I'm a college student working on a senior design project for FPGA's. I'm interested to know if Packet Sender has the ability to send a stream of packets continuously. If that is possible, the next question would be - is it possible to borrow this stream of packets from a secondary source? In my project, the most success achievable would be a demonstration of the FPGA accepting a continuous stream of pseudo-random bitstream data. I don't expect Packet Sender to natively have this function integrated, but I am interested in the ability to take data from a rand() function, or a PRBS compliant equation (such as https://blog.kurttomlinson.com/posts/prbs-pseudo-random-binary-sequence).

As a side note, I know this has to be said in open source - I am not asking you to actually code (I am fluent in C++) any of this, just investigating what capacities the software currently has, and if you could point me to the right direction in the code base. I'd be more than happy to push changes upstream if these are features that are not present and you think should be.

Thanks!

**Reply by dannagle** [reply]
You didn't specify UDP or TCP. TCP will be a bit harder to get Packet Sender features to what you need.

- TCP: Packet Sender has a {{RANDOM}} macro for sending a single random number. It also has persistent connections. There is an open bug to keep repeating and have macros work within a persistent window. Between that you are 80% there.
- UDP: There is already "intense traffic generator" tool that streams UDP to an IP and port. The only tweak would be to allow it support you special random data. Note that the intense traffic generator is best-effort and not super-accurate, but it may do what you need as-is. 
- UDP: Constantly repeating macros works on UDP (there is no persistent connection). You could probably do what you need with just having a new macro, perhaps {{BINARY_RANDOM}}. The speed of the repeating is not very high (thus the reason intense traffic generator was developed), but it may be enough.

**Comment by MadMatty101** [comment]
Hi, I was wondering if there is a way to clear the saved packets when the program is closed other than manually deleting them. The reason is that I will have multiple database files but with different number of saved files and with different names, which need to be sent just once before closing OR opening a new database. I see will overwrite existing ones of the same name. The only way I see at the moment is to have all with the same number of saved packets  with sequential names the same and just have blanks in the un-needed ones. 
Thanks Matty

**Reply by dannagle** [reply]
Not quite sure what you are trying to do, but it may be easier to use the portable version of Packet Sender. You could have several copies and make desktop shortcuts going to each executable.

If that won't work, you could make some kind of launcher script. Instead of wiping the database when it closes, you would wipe it just before launching. Packets are stored in `packets.ini`. For portable mode, that file is local to the exe. For installed versions, it uses `%APPDATA%`.

**Reply by MadMatty101** [reply]
I should have said I have no programming/scripting experience but I will give it a try (learning by trying). What I am trying to do long term is send multiple commands to multiple IP's (like inkjets / bar code readers) with the data being pulled from a database with a single key press.
I won't be doing any of that but in the short term I needed a way to send the strings to each to keep it working in a way that someone with even a little knowledge (like me) could still do. 

Thanks for the tip. Matty

**Comment by jttttav** [comment]
Hello Dan,

I am searching for a solution to controlling PTZ cameras with IP VISCA commands with keyboard shortcuts via autohotkey. I think I should be able to do this by having auto hot key paste the commands within the command prompt interface, the only problem is that i am having trouble opening the command prompt interface. Could you walk me through the process?

Any help would be greatly appreciated.

Thanks,
Justin

**Reply by dannagle** [reply]
@jttttav 
I assume you are using Windows. A command prompt-only binary is included using the ".com" file extension. Reference that directly if you are having problems.  See screenshot.

Some CLI examples can be found (near the bottom) in the documentation:
https://packetsender.com/documentation

![image](https://user-images.githubusercontent.com/1140573/115155401-702b3700-a045-11eb-9589-0a6f9a295931.png)



Also, the GUI was built to support autohotkey if that is easier. Send the GUI Ctrl+1, Ctrl+2, etc. depending on what you are trying to accomplish.

**Comment by aershey-git** [comment]
Hello Dan,

Could you please direct me to the link where its architecture is detailed?

Thanks!

**Reply by dannagle** [reply]
The documentation is located here: https://packetsender.com/documentation

**Comment by WilkyPrananta** [comment]
Hi Dan,

How to send data using packetsender DLL from an application? How to call DLL function to send this data? 
For example this command in CLI : packetsender -uxw 500 192.168.0.100 8882 ”02 00 06 00 46 03 00 00 00 4F 03”. 

Thanks,
Wilky

**Reply by dannagle** [reply]
Hello, and welcome to the forums.

Packet Sender does not supply a callable DLL. As a workaround, you can perform a system call and capture the output.

**Reply by WilkyPrananta** [reply]
Thank you for your confirmation. Packetsender is amazing apps.

**Comment by BGrandhi** [comment]
Hi,
I am using the packet sender app to send a udp message from a virtual adapter(ethernet 6) present on my PC. I can see in the arp table of the virtual adapter(ethernet 6) contains the IP address of the destination to which I am trying to send the message .
But still when I click send message from the packet sender it is transmitting it from another ethernet adapter (ethernet 2) instead of my virtual ethernet adapter (ethernet 6).

Can you please help a way to make sure that a message being sent uses my virtual adapter (ethernet 6) just for the given destination IP. FYI, the ethernet adapter (ethernet 2)  which is currently being used by the packet sender is my vpn adapter used to connect to my office server.

**Reply by dannagle** [reply]
If the UDP packets aren't going out the NIC that you want, you can try to bind the IP address of Packet Sender to an IP owned by that NIC. 

Go to:  File -> Settings -> Network. 

Look on the top right and type in the IP of the NIC. 

See screenshot:

![image](https://user-images.githubusercontent.com/1140573/236591810-6f5d7b0c-bf12-49cf-993b-56f538e2da23.png)

Packet Sender will bind the main UDP socket to that IP. The packets should then flow out that NIC.

**Reply by BGrandhi** [reply]
Hi Dan,
Thank you so much for the response. I have tried your suggestion and unfortunately it wasn't working in my case. Once I bind it to the IP address of required adapter(ethernet 2) then I dont see any message being sent out in any of the Ethernet adapters available. 
Once I remove the config for binding the IP address then I can observe the previous behavior where the packet is being sent through my virtual adapter (ethernet 6) instead of my expectation of ethernet 2 adapter. 
I have tried attaching a screenshot for better explanation but it gives out error saying can't process the file which is of png type. So I have added a pdf file

[Capture1.pdf](https://github.com/dannagle/PacketSender/files/11421331/Capture1.pdf)

Can you suggest if I can try anything else for debugging? FYI, I am using some python scripts using scapy to send out some Ethernet messages and communicate using virtual ethernet 6 adapter which is working fine.

**Comment by dannagle** [comment]
Nothing is jumping out at me with the screen capture.

The next level of network troubleshooting is to disable all other ethernet adapters except the one you want to test with. Disable all but the one you want and set Packet Sender back to general IPv4 bind. The packet has no choice but to go out that adapter. Give that a go.

**Reply by BGrandhi** [reply]
Hi Dan,

I appreciate your quick responses. Yeah if I do that then I can see it is going through the virtual ethernet 6 adapter which I was expecting. But I cannot proceed like that as I need the ethernet 2 adapter enabled to have the other tools in my use case working as they are running on server licenses. 
So I have to figure out a way to send the pack through virtual ethernet 6 adapter even when ethernet 2 adapter enabled.

**Comment by dannagle** [comment]
You might have a soft IP conflict.  This is a conflict where the IP is not directly conflicting, but they overlap subnets. 

What are the IP addresses and subnets of the 2 adapters?  The one you want to use and the one it is currently sending out of?

**Reply by BGrandhi** [reply]
The one I want to use has an IP address of 192.168.1.2 with a subnet mask of 255.255.255.0. The one which is currently sending out the message has an IP address of 10.252.244.24 with a subnet mask of 255.255.192.0. 

All the time from packet sender,I am trying to send a message to an IP address of  192.168.1.12

**Comment by dannagle** [comment]
Maybe you need to clear your arp table. Perhaps something else responded as having 192.168.1.12.

Launch admin-level powershell and run `arp -d` then try again using Packet Sender force-bind to 192.168.1.2.

**Reply by BGrandhi** [reply]
Hi Dan, 
Yes I have tried that. Its still the same. Just after deleting the arp table, If I bind the IP to 192.168.1.2 then I cannot see the message being sent out from any of the ethernet adapters. Once I remove the binding then it is being sent through Ethernet 2 adapter which is a VPN network  I am connected to (has an IP address of 10.252.244.24 with a subnet mask of 255.255.192.0) like before .

At this point, I think this behavior is may be because of how my VPN is setup to make it route everything through it.

**Comment by Adevill20** [comment]
Hi Dan
I'm using the software to send UDP packets through a data diode (one way traffic) over a period of 24 hours, which is working fine. The problem is the log window only holds up 10001 lines of logs. Is there a way to log directly to a file when sending/receiving packets?

**Reply by dannagle** [reply]
That has been requested before. It just has not been a priority. Add your comment here so that gets more visibility:  #191

**Comment by deepthideepthi-sag** [comment]
Hi Dan,

Is there a slim, cli only version available? We work with "resource-constrained" linux devices and was wondering if something like that is already available and useable

**Reply by dannagle** [reply]
It's not much, but maybe it'd be enough, but you can compile from source a command-line only version of Packet Sender. You will still require the core Qt libraries, but you will at least avoid needing any GUI stuff.

The Qt project file for CLI-only is here:
https://github.com/dannagle/PacketSender/blob/v8.5.2/src/PacketSenderCLI.pro

It sets the project to console mode and removes all links to GUI widgets.

**Comment by pandesandip062** [comment]
Hi dannagle,

I am using packet sender for one of my project and we are going automate packet sender through CMD but unfortunately am getting the same response.


![Uploading cmd.png…]()

**Reply by dannagle** [reply]
Were you trying to upload a screenshot? I can't see it. Will you try uploading again?

**Reply by pandesandip062** [reply]
Sure
<img width="857" alt="cmd" src="https://github.com/dannagle/PacketSender/assets/128965482/c41b7ba1-3a4e-4f10-9c4e-7089cd70a716">

**Comment by fabulousdeepaka** [comment]
Hi dannagle, 

I hope to recieve quick response from you😊

my question is, Can I use this packet sender to generate fake traffic into my website???🙃

**Reply by dannagle** [reply]
Yes, it can, though it is fairly limited compared to more dedicated tools for that kind of task. 

You can do HTTP(S) calls and add a resend delay. Here is send a GET request to example.com every second...

![image](https://github.com/dannagle/PacketSender/assets/1140573/aaa45491-e1f7-4a4b-afc2-69c68b26efe4)


Packet Sender supports creating custom GET and POST requests. You can keep clicking Send to stack the resend requests.  Perhaps that is enough to accomplish your goal.

**Comment by eleliope** [comment]
Hi Dan,
I'm eleliope, I'm part of a small team of 4 people dedicated to lighting and audiovisuals. We found PacketSender, and we are finding it a very good tool. We are testing it to send PJLink signals for projector control, UDP signals for mediaplayer control, and TCP signals for a DMX controller (LanBox). 
We have a question regarding the possibility of launching in panel-only mode using the command line option --starterpanel. I write it in a discussion that is already open on this topic (Direct Display Pannel #301).
We would like to tell you that, in addition to being easy and effective to use, the information contained in the documentation is very clear and useful.

**Reply by dannagle** [reply]
Hello eleliope, I'm glad you like Packet Sender. I will answer your question in the other thread.

**Comment by H4rdM5t1zl0V** [comment]
Greetings Dan,
Recently started working in lab for testing internet switches and we needed a traffic generator. Most of our staff are Windows-based people, so the idea of Cisco Trex was met with unhappy expressions =(
We needed a traffic generator to send tagged and untagged traffic through switch from one network card port to the other. I showed Packet Sender to my boss and he said that it looks great, so i was hoping if you can tell me if it is possible or do we need to setup a second pc as a receiver with Wireshark on the other end instead ? 
Thank you in advance
-Mstislav

**Reply by dannagle** [reply]
VLAN tagging happens at Layer 2. Packet Sender operates at Layer 3 and there is no plans to give it lower level access. One of the guiding design principles of Packet Sender is for all features to be available without admin permissions. This is what makes the purely portable version possible. Connecting a simple managed switch would likely have the feature you need and do not cost much.

**Comment by briann5545** [comment]
Why is Packet Sender opening a new log window every time it receives a small group of packets?  And why are these log windows white on white text/background?  There are no settings for any of this
![PacketSender](https://github.com/dannagle/PacketSender/assets/45271951/54aa09ee-1062-42f9-8fa7-56bc0c2edbb7)
...

**Comment by dannagle** [comment]
Uncheck "persistent TCP" on the main window.  

![image](https://github.com/dannagle/PacketSender/assets/1140573/0a3c2a60-ef6b-43e5-9869-33155ab5963e)


The log window is white because there is a race condition with the coloring when the connection closes quickly. There is an open bug report on that.

**Comment by kiranhosur-dev** [comment]
Working on FHIR. I want to send FHIR XML String as inbound by TCP to a specific port number. How to send fhir (XML) string by TCP?
![A02](https://github.com/dannagle/PacketSender/assets/78899667/ce43bfee-e952-47d4-be82-c6e37d8bb7aa)

**Reply by dannagle** [reply]
You can double-click the ASCII line editor to bring up multi-line and paste in your XML text directly. You can also load in the file using the "Load File" button.

**Comment by GauthamShankar21** [comment]
Hey Dan, 
I just wanted to know can I work with the TCP Echo server with this application? I tried to do but I have not received the data back other than the ACK, Please let me know what configs I need to make?

**Comment by dannagle** [comment]
There is not a direct echo feature. You can have it reply with macros.

File->Settings->Network, then look bottom left.

![image](https://github.com/dannagle/PacketSender/assets/1140573/799ae6c7-6798-472f-98e5-d5c21d8d4a83)

That will reply back with the current time when connected to on TCP/UDP/SSL.

Packet Sender supports several macros you can piece together to try to do what you need. Hope that helps.

**Comment by ralongi** [comment]
Hi Dan,

I'd like to use the Packet Sender CLI to run some automated tests/scripts.  One of the requirements is to generate random packet sizes.  Is this possible using the CLI?

Thanks!

**Reply by dannagle** [reply]
There is not one built-in, but it'd be pretty easy to do via scripting. 

Here is an example using Powershell. Save this to a file (like `random-data-packetsender.ps1` and then run it using `./random-data-packetsender.ps1`

```powershell
# Generate a random length for the hex string between 10 and 500
$hexLength = Get-Random -Minimum 10 -Maximum 501  # 501 to include 500 as max

# Define a valid hex character set
$hexChars = @('0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F')

# Generate a random hex string of the required length using ForEach-Object
$hexData = -join (1..$hexLength | ForEach-Object { Get-Random -InputObject $hexChars })

# Display the generated hex data
Write-Host "Generated Hex Data: $hexData"

# Path to Packet Sender executable 
$packetSenderPath = "C:\Program Files\PacketSender\packetsender.com"

# Arguments for Packet Sender
$arguments = @(
    "-txw", "500",            # TCP, hex, wait 500 ms
    "127.0.0.1",              # Destination address
    "49697",                  # Port
    $hexData                  # Hex data argument
)

# Run Packet Sender with the generated hex data and additional arguments
Start-Process -FilePath $packetSenderPath -ArgumentList $arguments
```

**Comment by tblount70** [comment]
I could use a little help understanding the Load File feature. Specifically, I cannot seem to create a text file that will load a packet in the format I want.

The screenshot below shows saved packets I want to send (these work well). I have a separate application that will be generating larger packets (170 bytes) with data that varies. My goal is to use the Load File button to bring those packets into PacketSender. In no case will a saved txt file include more than one packet.

The screenshot shows what happened after this sequence:
1) I double-clicked the ASCII field of saved packet 1. Then I copied that data.
2) I pasted the data into a txt file.
3) I used the Load button to bring the packet in for immediate use.

It seems that anything Load File brings in is interpreted literally as ASCII. Short of generating the txt files with non-readable control characters, is there any way to get Load File to read in as raw hex? I've googled a bit, and I've tried several things. Have not discovered the magic yet... 

<img width="914" height="782" alt="image" src="https://github.com/user-attachments/assets/e9144574-8554-4895-8b79-3313d6fccdbf" />

**Reply by dannagle** [reply]
@tblount70 ,  

You have binary data in your packets. It is the HEX field that you want.  That is the raw bytes that actually get sent out.  The ASCII field in Packet Sender is actually a mixed ASCII notation using some short codes for non-printable characters, like `\00` to represent NULL or `\r` to represent a carriage return. 

There isn't an export binary function in Packet Sender. The intent was for the ASCII to help with that, but ASCII will not work (as you mentioned).

Fortunately, getting that data into a file is simple in Linux or Mac terminal. WSL or Git Bash on Windows could do this too:
```bash
# Copy your hex over here...
echo "48 65 6c 6c 6f" | xxd -r -p > mypacket.bin
```
That will create mypacket.bin that represents your packet in a binary file that your other program can read in. 


If you don't have a Linux or Mac prompt, I am not skilled in Powershell, but I think this will work for it...
```powershell
# Copy your hex over here...
$hex = "48 65 6c 6c 6f"

# This will remove the space delineations
$cleanHex = $hex -replace '\s|-',''

$bytes = for($i=0; $i -lt $cleanHex.Length; $i+=2){ 
    [Convert]::ToByte($cleanHex.Substring($i, 2), 16) 
}
[System.IO.File]::WriteAllBytes("mypacket.bin", $bytes)

```

Hope that helps.

**Reply by tblount70** [reply]
Thank you!

**Comment by kristakis** [comment]
Hi,
I have set PacketSender up to reply to a call I am making from my software. I would like to add a 2nd reply to the call a few seconds later. Is there a way to do this? I tried using the sending feature but that is a separate connection and not a reply on the incoming connection.
thanks,
Kris

**Reply by dannagle** [reply]
Hello, there is no built-in way to do that. There isn't anything that is "wait X seconds and send additional data" mechanism. Sending starts a new connection (as you found). 

There is a "smart reply" mechanism that is close to what you are wanting, but that only triggers on incoming data. If the other side sends a response, you could trigger on that with a smart reply. That reuses the session.
